client/grpc_query.go

```diff
@@ -72,35 +68,47 @@ func (ctx Context) Invoke(grpcCtx gocontext.Context, method string, req, reply i
 		ctx = ctx.WithHeight(height)
 	}
 
-	abciReq := abci.RequestQuery{
-		Path:   method,
-		Data:   reqBz,
-		Height: ctx.Height,
-	}
+	if ctx.ClientConn != nil {
+		err = ctx.ClientConn.Invoke(grpcCtx, method, req, reply, opts...)
```

Comment:
for https://github.com/cosmos/cosmos-sdk/pull/10997#issuecomment-1022417957

we might need:

```suggestion
		if ctx.Height != nil {
			grpcCtx := metadata.AppendToOutgoingContext(grpcCtx, grpctypes.GRPCBlockHeightHeader, ctx.Height)
		}
		err = ctx.ClientConn.Invoke(grpcCtx, method, req, reply, opts...)
```

The reason is:
- before: if there was a incoming request header set, then we set `ctx.WithHeight(read from header)`
- now: if ctx.Height is set, then we add a grpc header. 

Note: If both grpcCtx already has a header metadata, and ctx.Height is not nil, I think the grpcCtx metadata header should take precedence

Reply:
i'm not sure `grpcCtx` will have any metadata set prior call to `err = ctx.ClientConn.Invoke`

also append to your suggestion 
```go
		if ctx.Height > 0 {
			grpcCtx = metadata.AppendToOutgoingContext(grpcCtx, grpctypes.GRPCBlockHeightHeader, strconv.FormatUint(uint64(ctx.Height), 10))
		}
```